File: 1001-2000/1650. Lowest Common Ancestor of a Binary Tree III.java
// Medium
// Set
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/lowest-common-ancestor-of-a-binary-tree-iii/

import java.util.*;
class Solution {
    // Since there is no specific root node, but we can track its parent node 
    // Traverse all ancestors of 'p' and store in set, and traverse ancestors of 'q', the first overlap is LCA
    public Node lowestCommonAncestor(Node p, Node q) {
        Set<Node> set = new HashSet<>();
        for (Node node = p; node != null; node = node.parent) {
            set.add(node);
        }
        for (Node node = q; node != null; node = node.parent) {
            if (!set.add(node)) {
                return node;
            }
        }
        return null;
    }
}

File: 1001-2000/1650. Lowest Common Ancestor of a Binary Tree III.py
# Time:O(h), Space:O(h)

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None

class Solution:
    def lowestCommonAncestor(self, p: "Node", q: "Node") -> "Node":
        anc = set()

        cur = p
        while cur:
            anc.add(cur)
            cur = cur.parent

        cur = q
        while cur:
            if cur in anc:
                return cur
            cur = cur.parent
